packages/core/src/lib/filter/v2/index.ts

```diff
@@ -0,0 +1,334 @@
+import type { Libp2p } from "@libp2p/interface-libp2p";
+import type { Peer } from "@libp2p/interface-peer-store";
+import { IncomingStreamData } from "@libp2p/interface-registrar";
+import type {
+  ActiveSubscriptions,
+  Callback,
+  IDecodedMessage,
+  IDecoder,
+  IFilterV2,
+  IProtoMessage,
+  ProtocolCreateOptions,
+  ProtocolOptions,
+} from "@waku/interfaces";
+import { WakuMessage } from "@waku/proto";
+import debug from "debug";
+import all from "it-all";
+import * as lp from "it-length-prefixed";
+import { pipe } from "it-pipe";
+
+import { BaseProtocol } from "../../base_protocol.js";
+import { DefaultPubSubTopic } from "../../constants.js";
+import { groupByContentTopic } from "../../group_by.js";
+
+import {
+  FilterPushRpc,
+  FilterSubscribeResponse,
+  FilterSubscribeRpc,
+} from "./filter_rpc.js";
+
+const log = debug("waku:filter:v2");
+
+export type UnsubscribeFunction = () => Promise<void>;
+export type RequestID = string;
+
+type Subscription<T extends IDecodedMessage> = {
+  decoders: IDecoder<T>[];
+  callback: Callback<T>;
+  pubSubTopic: string;
+};
+
+const FilterV2Codecs = {
+  SUBSCRIBE: "/vac/waku/filter-subscribe/2.0.0-beta1",
+  PUSH: "/vac/waku/filter-push/2.0.0-beta1",
+};
+
+/**
+ * Implements client side of the [Waku v2 Filter protocol](https://rfc.vac.dev/spec/12/).
+ *
+ * Note this currently only works in NodeJS when the Waku node is listening on a port, see:
+ * - https://github.com/status-im/go-waku/issues/245
+ * - https://github.com/status-im/nwaku/issues/948
+ */
+class FilterV2 extends BaseProtocol implements IFilterV2 {
+  options: ProtocolCreateOptions;
+  private subscriptions: Map<RequestID, unknown>;
```

Comment:
I don't think this is necessary for filter v2. a `request id` is only for one response-reply and should not be re-used between a subscribe and an unsubscribe. see https://github.com/waku-org/js-waku/pull/1302/files#r1178541406

Reply:
Indeed. No use to persist request ID state. In fact, don't re-use request IDs, as this will make specific ACKs ambiguous.